packages/flutter/lib/src/widgets/splash.dart

```diff
@@ -0,0 +1,621 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'package:flutter/foundation.dart';
+import 'package:flutter/rendering.dart';
+
+import 'framework.dart';
+import 'lookup_boundary.dart';
+import 'notification_listener.dart';
+import 'ticker_provider.dart';
+
+// Examples can assume:
+// late Color color;
+// late BuildContext context;
+
+/// Signature for the callback used by [Splash]es to obtain the
+/// appropriate [Rect].
+typedef RectCallback = Rect Function();
+
+/// A splash of [color] shown in response to a user gesture.
+///
+/// The [confirm] and [cancel] methods alter the splash based on the
+/// user's behavior. For example, a tap event might trigger a splash
+/// that's confirmed when the corresponding [GestureDetector.onTapUp]
+/// event is seen and canceled when the pointer is dragged out of the
+/// reference box.
+///
+/// See also:
+///
+/// * [SplashBox], the widget that paints splashes.
+/// * [SplashFactory], a class that creates splashes using a specific
+///   function signature.
+/// * [InkResponse], which creates splashes using a [SplashFactory].
+abstract class Splash {
+  /// Creates an interactive splash effect.
+  Splash({
+    required this.controller,
+    required this.referenceBox,
+    Color color = const Color(0x00000000),
+    ShapeBorder? customBorder,
```

Comment:
I think the `Splash` class should be generalized enough so that it
* Does not assume a solid color (it might draw a gradient)
* Does not assume a shape (the `customBorder` doesn't assume, but why does it have to be here anyway)
* And does not assume the `RenderBox` protocol, but the more general `RenderObject` protocol

Reply:
> Does not assume a solid color (it might draw a gradient)

That makes sense. I'd be in favor of making the `color` nullable, although I believe it could be a breaking change and should probably be saved for a future PR.

<br>

> Does not assume a shape (the `customBorder` doesn't assume, but why does it have to be here anyway)

The `customBorder` field is intended to optionally provide a path for the splash to be clipped to.

<br>

> And does not assume the `RenderBox` protocol, but the more general `RenderObject` protocol

(see [other thread](https://github.com/flutter/flutter/pull/149963/files#r1672932488))

Reply:
I think it's better to create a subclass that assumes the splash takes a `ShapeDecoration`, `ShapeBorder`, and an optional shaped clip. The general class should be customizable to draw any animation.

Reply:
That sounds fantastic!

<br>

Maybe the new subclass could look something like this:

```dart
abstract class DecorationSplash extends Splash {
  DecorationSplash({
    required ShapeDecoration decoration,
    required super.controller,
    required super.referenceBox,
    super.onRemoved,
  }) : super(
         color: decoration.color,
         customBorder: decoration.shape,
       );
}
```

<br>

> The general class should be customizable to draw any animation.

Yeah that makes sense, and I think it does a pretty good job of that already—in my opinion, configuring a `Splash` using a single color is common enough that it makes sense for the general class to have a `color` member.

But after taking another look, I see that the class is already using the "private value, public getter" pattern, so we could make it nullable right now without any breakages!

I'll go ahead and do that now; let me know if there are any other changes you'd like to see in this PR.

Reply:
This was a super satisfying change.

```dart
// before
abstract class InkFeature extends Splash {
  /// Initializes fields for subclasses.
  @Deprecated(
    'Use Splash instead. '
    'Splash effects no longer rely on a MaterialInkController. '
    'This feature was deprecated after v3.23.0-0.1.pre.',
  )
  InkFeature({
    required super.controller,
    required super.referenceBox,
    super.color = Colors.transparent,
    super.onRemoved,
  });
}
typedef InteractiveInkFeature = Splash;
```

```dart
// after
typedef InkFeature = Splash;
typedef InteractiveInkFeature = Splash;
```

Thanks again for the advice! I'll go ahead and mark this as "resolved"; feel free to unresolve or start a new thread if there's something else you'd like to see.